src/scipp/testing/strategies.py

```diff
@@ -0,0 +1,225 @@
+    # except NULL, which causes problems in C and C++ code (e.g. HDF5).
+    return st.text(st.characters(
+        whitelist_categories=['L', 'M', 'N', 'P', 'S', 'Zs', 'Cc'],
+        blacklist_characters='\0'),
+                   min_size=0,
+                   max_size=10)
+
+
+def sizes_dicts(
+        ndim: Optional[Union[int, st.SearchStrategy]] = None) -> st.SearchStrategy:
+    if isinstance(ndim, st.SearchStrategy):
+        return ndim.flatmap(lambda n: sizes_dicts(ndim=n))
+    keys = dims()
+    values = st.integers(min_value=1, max_value=10)
+    if ndim is None:
+        # The constructor of sc.Variable in Python only supports
+        # arrays with <= 4 dimensions.
+        return st.dictionaries(keys=keys, values=values, min_size=0, max_size=4)
+    return st.dictionaries(keys=keys, values=values, min_size=ndim, max_size=ndim)
+
+
+def units() -> st.SearchStrategy:
+    return st.sampled_from(('one', 'm', 'kg', 's', 'A', 'K', 'count'))
+
+
+def integer_dtypes(sizes: Sequence[int] = (32, 64)) -> st.SearchStrategy:
+    return st.sampled_from([f'int{size}' for size in sizes])
+
+
+def floating_dtypes(sizes: Sequence[int] = (32, 64)) -> st.SearchStrategy:
+    return st.sampled_from([f'float{size}' for size in sizes])
+
+
+def scalar_numeric_dtypes() -> st.SearchStrategy:
+    return st.sampled_from((integer_dtypes, floating_dtypes)).flatmap(lambda f: f())
+
+
+def use_variances(dtype) -> st.SearchStrategy:
+    if dtype in (DType.float32, DType.float64):
+        return st.booleans()
+    return st.just(False)
+
+
+def _variables_from_fixed_args(args) -> st.SearchStrategy:
+    def make_array():
+        return npst.arrays(args['dtype'],
+                           tuple(args['sizes'].values()),
+                           elements=args['elements'],
+                           fill=args['fill'],
+                           unique=args['unique'])
+
+    return st.builds(partial(creation.array,
+                             dims=list(args['sizes'].keys()),
+                             unit=args['unit']),
+                     values=make_array(),
+                     variances=make_array() if args['with_variances'] else st.none())
+
+
+@st.composite
+def variable_args(draw,
+                  *,
+                  ndim=None,
+                  sizes=None,
+                  unit=None,
+                  dtype=None,
+                  with_variances=None,
+                  elements=None,
+                  fill=None,
+                  unique=None) -> dict:
+    if ndim is not None:
+        if sizes is not None:
+            raise InvalidArgument('Arguments `ndim` and `sizes` cannot both be used. '
+                                  f'Got ndim={ndim}, sizes={sizes}.')
+    if sizes is None:
+        sizes = sizes_dicts(ndim)
+    if isinstance(sizes, st.SearchStrategy):
+        sizes = draw(sizes)
+
+    if unit is None:
+        unit = units()
+    if isinstance(unit, st.SearchStrategy):
+        unit = draw(unit)
```

Comment:
How does hypothesis deal with this, conceptually? Does it search these things independently? *We* know that errors/bugs from unit handling are (with high likely hood?), e.g., independent of errors/bugs from, i.e., searching each "dimension" (I mean the units and sizes dimensions, not scipp dimensions) independently is wasteful. Does this make things slow? Will it limit us in the case of DataArray, since that has even more params?

Is there a way to use this knowledge?

Similar example: Bugs relating to dim names are unrelated to bugs from dim sizes, so a naive independent search for 1-D variables might need to look at:
```
x: 0
x: 1
x: 2
y: 0
y: 1
y: 2
z: 0
...
...
```
while an informed one could just use:
```
x: 0
y: 1
z: 2
```
i.e., cover both dimensions of the parameter space that the same time.

Reply:
> Is there a way to use this knowledge?

Good question!

In my understanding, hypothesis does not correlate sampling from component strategies in this way. There is a way to guide sampling: https://hypothesis.readthedocs.io/en/latest/details.html?highlight=target maybe this can be used to do what you say.

> Does this make things slow?

Only shrinking of failed examples because the parameter space is large. But as long as there are no failures, it always generates the same number of examples regardless of how many parameters there are. But of course, this means that coverage of parameter space is not optimal.

Reply:
> But as long as there are no failures, it always generates the same number of examples regardless of how many parameters there are. But of course, this means that coverage of parameter space is not optimal.

What I suggest above effectively means exploring 2 dimensions of the parameter space at once? I suppose this would improve coverage, since it effectively gets us from N**2 to N?

Just thinking, not suggesting that this is something we should do (or can do).

Reply:
Might be worth keeping in mind if and when we add more tests. But we have to be very careful not to miss relevant cases when restricting parameterspace.